File: hackerrank/30-days-of-code/inheritance/inheritance.cpp
#include <iostream>
#include <vector>

using namespace std;


class Person{
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification){
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout<< "Name: "<< lastName << ", "<< firstName <<"\nID: "<< id << "\n"; 
		}
	
};

class Student :  public Person{
	private:
		vector<int> testScores;  
	public:
        Student(string firstName, string lastName, int identification, vector<int> scores) : Person(firstName, lastName, identification) {
			this->testScores = scores;
		}

		char calculate() {
			int average = 0;
			for (int i = 0; i < testScores.size(); i++)
			{
				average += testScores[i];
			}
			average = average/testScores.size();

			if (average >= 90 && average <= 100) {
				return 'O';
			}
			else if (average >= 80 && average < 90) {
				return 'E';
			}
			else if (average >= 70 && average < 80) {
				return 'A';
			}
			else if (average >= 55 && average < 70) {
				return 'P';
			}
			else if (average >= 40 && average < 55) {
				return 'D';
			}
			else {
				return 'T';
			}
			
		}
};

int main() {
	string firstName;
  	string lastName;
	int id;
  	int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  	vector<int> scores;
  	for(int i = 0; i < numScores; i++){
	  	int tmpScore;
	  	cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}

File: hackerrank/30-days-of-code/inheritance/inheritance.py
#!/usr/bin/env python

class Person:
	def __init__(self, firstName, lastName, idNumber):
		self.firstName = firstName
		self.lastName = lastName
		self.idNumber = idNumber

	def printPerson(self):
		print("Name:", self.lastName + ",", self.firstName)
		print("ID:", self.idNumber)

class Student(Person):
    #   Class Constructor
    
    #   Parameters:
    #   firstName - A string denoting the Person's first name.
    #   lastName - A string denoting the Person's last name.
    #   id - An integer denoting the Person's ID number.
    #   scores - An array of integers denoting the Person's test scores.
    #
    # Write your constructor here
    def __init__(self, firstName, lastName, idNumber, scores):
        self.firstName = firstName
        self.lastName = lastName
        self.idNumber = idNumber
        self.scores = []

    #   Function Name: calculate
    #   Return: A character denoting the grade.
    #
    # Write your function here
    def calculate(self):
        sum = 0
        for i in range(len(scores)):
            sum += scores[i]
        average = sum/len(scores)
        if average >= 90 and average <= 100:
            grade = "O"
        elif average < 90 and average >= 80:
            grade = "E"
        elif average < 80 and average >= 70:
            grade = "A"
        elif average < 70 and average >= 55:
            grade = "P"
        elif average < 55 and average >= 40:
            grade = "D"
        else:
            grade = "T"
        return grade        
    
line = input().split()
firstName = line[0]
lastName = line[1]
idNum = line[2]
numScores = int(input()) # not needed for Python
scores = list(map(int, input().split()))
s = Student(firstName, lastName, idNum, scores)
s.printPerson()
print("Grade:", s.calculate())
